### PR TITLE
[CBRD-21781] Memory Leaks from string_buffer

### PR DESCRIPTION
**Pull request by bsolomenco**
fixed memory leaks and mismatched new/delete

also:
http://jira.cubrid.org/browse/CBRD-21785
http://jira.cubrid.org/browse/CBRD-21778
http://jira.cubrid.org/browse/CBRD-21777
http://jira.cubrid.org/browse/CBRD-21765
http://jira.cubrid.org/browse/CBRD-21772

no ideea why sql/_08_javasp/cases/413-2.sql fails in CI, it runs OK locally  and it doesn't appear to have anything to do with my changes

### CODE REVIEW

File: src/base/mem_block.hpp [hunk 1]
```diff
@@ -70,6 +70,12 @@ namespace mem
 	, ptr { (char *)ptr}
       {}
 
+      virtual ~block()
```
Comment (ribram) [inline src/base/mem_block.hpp]: Can you please explain the virtual destructor? Do we delete mem::block pointers to derived instances?

File: src/base/string_buffer.hpp [hunk 2]
```diff
@@ -91,7 +91,7 @@ class string_buffer: public mem::block_ext //collect formatted text (printf-like
 
     void add_bytes (size_t len, void *bytes);                     //add "len" bytes (can have '\0' in the middle)
 
-    template<typename... Args> inline void operator() (Args &&... args); //add with printf format
+    template<typename... Args> inline int operator() (Args &&... args); //add with printf format
```
Comment (ribram) [inline src/base/string_buffer.hpp]: Is return value ever used? Please explain the change.

File: src/base/string_buffer.hpp [hunk 3]
```diff
@@ -108,13 +108,21 @@ void string_buffer::operator+= (const char ch)
 {
   if (dim < m_len + 2)
```
Comment (ribram) [inline src/base/string_buffer.hpp]: Why "dim < m_len + 2" and not "dim <= m_len"? The "2" seems to be required only if dim == 0.

File: src/base/mem_block.hpp (hunk 1 above)
Comment (bsolomenco) [reply src/base/mem_block.hpp]: currently no, but technically it is possible

File: src/base/string_buffer.hpp (hunk 2 above)
Comment (bsolomenco) [reply src/base/string_buffer.hpp]: the return value is not used yet but reviewing the code I found places where this value could be useful, eliminating the need of additional buffers; I put it there for future use

File: src/base/mem_block.hpp (hunk 1 above)
Comment (ribram) [reply src/base/mem_block.hpp]: I'd like to suggest hiding the destructor under protected access until we really need it to be virtual.

File: src/base/string_buffer.hpp (hunk 2 above)
Comment (ribram) [reply src/base/string_buffer.hpp]: Can you show me an example?

File: src/base/string_buffer.hpp (hunk 3 above)
Comment (bsolomenco) [reply src/base/string_buffer.hpp]: m_len doesn't count the ending '\0'
for the case when dim != 0 and dim == m_len+1 and I want to add another char then I have to realloc => condition to realloc is dim <= m_len+1 <=> dim < m_len+2
I'l rewrite the conditions as:
if(dim == 0)
  extend(2);
else if(dim <= m_len+1)
  extend(1);

Comment (ribram) [reply src/base/string_buffer.hpp]: Ok, looks good. Please add a comment to mention that m_len does not include '\0'.

File: src/base/string_buffer.hpp (hunk 2 above)
Comment (bsolomenco) [reply src/base/string_buffer.hpp]: db_value_printer.cpp: describe_real()
if we know how many chars have been added then we can revert it in place and add a different text

File: src/base/mem_block.hpp (hunk 1 above)
Comment (bsolomenco) [reply src/base/mem_block.hpp]: protected: block::~block(); => impossible to use instances of this class outside hierarchy
... and yet it is used in default/private_realloc/dealloc()
There is a workaround for the above particular cases but it is undesirable: to unwrap the code manually each time

File: src/base/string_buffer.hpp (hunk 2 above)
Comment (ribram) [reply src/base/string_buffer.hpp]: ok

**Pull request merged**